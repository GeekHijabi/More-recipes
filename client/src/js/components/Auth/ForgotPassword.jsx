import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import RecipeHeader from '../Partials/RecipeHeader';
import { apiForgotPassword } from '../../actions/auth';
import
validateforgotPasswordInput from
  '../../utils/validations/forgotPasswordValidation';

/**
 * @class ForgotPassword
 * @extends {React.Component}
 */
export class ForgotPassword extends React.Component {
  /**
   * @description Constructor function
   * @param {any} props
   * @memberof SignIn
   * @return {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  /**
   *
   * @param {any} event
   * @memberof ForgotPassword
   * @returns {void}
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   *
   * @param {any} event
   * @memberof ForgotPassword
   * @returns {void}
 */
  onClick(event) {
    if (this.isValid()) {
      // const url = window.location.href;
      // const currentURL = url.split('/')[url.split('/').length - 1];
      // console.log('cr', currentUrl);
      event.preventDefault();
      this.props.apiForgotPassword(2)
        .then((res) => {
          toastr.options = {
            closeButton: true,
            progressBar: true
          };
          this.props.history.push('/reset-password');
          toastr.success(res.data.message);
        })
        .catch((err) => {
          toastr.danger(err.data);
        });
      this.setState({ email: '' });
    }
  }

  /**
  *
  * @param {any} event
  * @memberof SignUp
  * @returns {object} event
  */
  isValid() {
    const { errors, isValid } = validateforgotPasswordInput(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
   * @description Constructor function
   * @param {any} props
   * @memberof ForgotPassword
   * @return {void}
   */
  render() {
    const { errors } = this.state;
    return (
      <div page="signin">
        <div className="overlay" />
        <RecipeHeader />
        <div className="card card-style">
          <div className="card-body mx-4">
            <div className="text-center">
              <h3 className="dark-grey-text mb-5" >
                <strong>Forgot password?</strong>
              </h3>
            </div>
            <div className="md-form">
              <label htmlFor="Form-email1" className="form-label">Email
            <input
              type="text"
              id="Form-email1"
              className="form-control"
              placeholder="Email"
              name="email"
              onChange={this.onChange}
              value={this.state.email}
              required
            />
                {errors.email &&
                <small style={{ color: '#A43741' }}>
                  {errors.email }
                </small>}
              </label>
            </div>
            <div className="text-center mb-3">
              <button
                type="button"
                className="btn btn-white btn-rounded"
                onClick={this.onClick}
              >
                       Send me an instruction
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
ForgotPassword.propTypes = {
  apiForgotPassword: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
};
export default connect(null, { apiForgotPassword })(ForgotPassword);
