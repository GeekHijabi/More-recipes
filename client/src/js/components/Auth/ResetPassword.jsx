import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import RecipeHeader from '../Partials/RecipeHeader';
import { apiResetPassword } from '../../actions/auth';
import
validateresetPasswordInput from
  '../../utils/validations/resetPasswordValidation';

/**
 * @class ResetPassword
 * @extends {React.Component}
 */
export class ResetPassword extends React.Component {
  /**
   * @description Constructor function
   * @param {any} props
   * @memberof SignIn
   * @return {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      newPassword: '',
      confirmPassword: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  /**
   *
   * @param {any} event
   * @memberof ResetPassword
   * @returns {void}
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   *
   * @param {any} event
   * @memberof ResetPassword
   * @returns {void}
 */
  onClick(event) {
    const baseUrl = window.location.href;
    const url = baseUrl.split('?token=');
    const currentURL = url[0];
    const userId = currentURL.split('/')[currentURL.split('/').length - 1];
    if (this.isValid()) {
      event.preventDefault();
      this.props.apiResetPassword(userId, this.state.newPassword, url[1])
        .then((data) => {
          toastr.options = {
            closeButton: true,
            progressBar: true
          };
          toastr.success(data.newPasswordMessage);
          this.props.history.push('/signin');
        })
        .catch((err) => {
          toastr.danger(err.data);
        });
    }
  }

  /**
  *
  * @param {any} event
  * @memberof ResetPassword
  * @returns {object} event
  */
  isValid() {
    const { errors, isValid } = validateresetPasswordInput(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
   * @description Constructor function
   * @param {any} props
   * @memberof ResetPassword
   * @return {void}
   */
  render() {
    const { errors } = this.state;
    return (
      <div page="signin">
        <div className="overlay" />
        <RecipeHeader />
        <div className="card card-style">
          <div className="card-body mx-4">
            <div className="text-center">
              <h3 className="dark-grey-text mb-5" >
                <strong>Reset password</strong>
              </h3>
            </div>
            <div className="md-form">
              <label htmlFor="Form-email1" className="form-label">
                <input
                  type="password"
                  className="Form-email1 form-control"
                  placeholder="New Password"
                  name="newPassword"
                  onChange={this.onChange}
                  value={this.state.newPassword}
                  required
                />
                {errors.password &&
                <small style={{ color: '#A43741' }}>
                  {errors.password }
                </small>}
              </label>

              <label htmlFor="Form-email1" className="form-label">
                <input
                  type="password"
                  className="Form-email1 form-control"
                  placeholder="confirm Password"
                  name="confirmPassword"
                  onChange={this.onChange}
                  value={this.state.confirmPassword}
                  required
                />
                {errors.confirmPassword &&
                <small style={{ color: '#A43741' }}>
                  {errors.confirmPassword }
                </small>}
              </label>
            </div>
            <div className="text-center mb-3">
              <button
                type="button"
                className="btn btn-white btn-rounded"
                onClick={this.onClick}
              >
                        Reset password
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
ResetPassword.propTypes = {
  apiResetPassword: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
};
export default connect(null, { apiResetPassword })(ResetPassword);
