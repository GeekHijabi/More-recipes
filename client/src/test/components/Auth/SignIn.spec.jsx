import React from 'react';
import { shallow } from 'enzyme';
import { SignIn } from '../../../js/components/Auth/SignIn';

const userInput = {
  identifier: 'tester',
  password: 'tester123'
};

const props = {
  apiLoginUser: jest.fn(() => Promise.resolve({ ok: true })),
  history: {
    push: () => {}
  }
};
describe('Component: SignIn', () => {
  beforeEach(() => {
    global.toastr = {
      success: () => {},
      error: () => {}
    };
  });

  describe('sign in snapshot', () => {
    it('should render correctly', () => {
      const signIn = shallow(<SignIn {...props} />);
      expect(signIn).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    it('should change event', () => {
      const event = {
        target: { name: 'password', value: 'tester123' },
      };
      const signIn = shallow(<SignIn {...props} />);
      const passwordField = signIn.find('#Form-pass1');
      passwordField.simulate('change', event);
      expect(signIn.instance().state.password).toBe('tester123');
    });
  });

  describe('onClick', () => {
    it('should login user when user details is set to the state', () => {
      const event = {
        preventDefault: jest.fn()
      };
      const submit = shallow(<SignIn {...props} />);
      submit.setState(userInput);
      submit.setState({ hasError: true, errorMessage: 'error', errors: {} });
      submit.instance().onClick(event);
    });

    it('should login error when incorrect detail is passed', () => {
      const event = {
        preventDefault: jest.fn()
      };
      props.apiLoginUser = () => Promise.reject(new Error('fail'));
      const submit = shallow(<SignIn {...props} />);
      submit.instance().onClick(event);
    });
  });

  describe('onDismiss', () => {
    it('should return state back to default', () => {
      const event = {
        preventDefault: jest.fn()
      };
      const dismiss = shallow(<SignIn {...props} />);
      dismiss.setState({ hasError: true, errorMessage: 'error' });
      const form = dismiss.find('.close');
      form.simulate('click', event);
    });
  });
});

