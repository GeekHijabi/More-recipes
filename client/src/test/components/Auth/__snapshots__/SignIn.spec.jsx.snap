// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SignIn sign in snapshot should render correctly 1`] = `
<div
  page="signin"
>
  <div
    className="overlay"
  />
  <div
    className="row"
  >
    <div
      className="col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <div
        className="card card-style"
      >
        <div
          className="card-body mx-4"
        >
          <div
            className="text-center"
          >
            <h3
              className="dark-grey-text mb-5"
            >
              <strong>
                Sign in
              </strong>
            </h3>
          </div>
          <div
            className="md-form"
          >
            <label
              className="form-label"
              htmlFor="Form-email1"
            >
              Username/Email
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 *
              </span>
              <input
                className="Form-email1 form-control required"
                name="identifier"
                onChange={[Function]}
                placeholder="username or Email"
                required={true}
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            className="md-form pb-3"
          >
            <label
              className="form-label"
              htmlFor="Form-pass1"
            >
              Password
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 *
              </span>
              <input
                className="form-control"
                id="Form-pass1"
                name="password"
                onChange={[Function]}
                pattern="(?=^.{8,15}$)(?!.*\\\\s).*$"
                placeholder="password"
                required={true}
                title="8 to 15 characters required"
                type="password"
                value=""
              />
              <p
                className="font-small blue-text d-flex justify-content-end"
              >
                <Link
                  className="blue-text ml-1"
                  href="/forgot-password"
                  replace={false}
                  to="/forgot-password"
                >
                  Forgot Password?
                </Link>
              </p>
            </label>
          </div>
          <div
            className="text-center mb-3"
          >
            <button
              className="btn btn-white btn-rounded submit float"
              onClick={[Function]}
              type="button"
            >
              Sign in
            </button>
          </div>
        </div>
        <div
          className="modal-footer mx-5 pt-3 mb-1"
        >
          <p
            className="font-small"
          >
            Yet to Register?
            <Link
              href="./signup"
              replace={false}
              to="/signup"
            >
               Sign Up
            </Link>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
